flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/StrategyUtils.java

```diff
@@ -0,0 +1,190 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.strategies;
+
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.CallContext;
+import org.apache.flink.table.types.logical.BinaryType;
+import org.apache.flink.table.types.logical.CharType;
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.VarBinaryType;
+import org.apache.flink.table.types.logical.VarCharType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeCasts;
+
+import javax.annotation.Nullable;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.apache.flink.table.types.logical.LogicalTypeFamily.APPROXIMATE_NUMERIC;
+import static org.apache.flink.table.types.logical.LogicalTypeFamily.EXACT_NUMERIC;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.BINARY;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.CHAR;
+import static org.apache.flink.table.types.logical.LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeCasts.supportsImplicitCast;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getLength;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getPrecision;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getScale;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasFamily;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasRoot;
+
+/**
+ * Utilities for shared logic in classes of this package.
+ */
+final class StrategyUtils {
+
+	/**
+	 * Finds a data type that is close to the given data type in terms of nullability and conversion
+	 * class but of the given logical root.
+	 */
+	static Optional<DataType> findDataType(
+			CallContext callContext,
+			boolean throwOnFailure,
+			DataType actualDataType,
+			LogicalTypeRoot expectedRoot,
+			@Nullable Boolean expectedNullability) {
+		final LogicalType actualType = actualDataType.getLogicalType();
+		return Optional.ofNullable(findDataTypeOfRoot(actualDataType, expectedRoot))
+			// set nullability
+			.map(newDataType -> {
+				if (Objects.equals(expectedNullability, Boolean.TRUE)) {
+					return newDataType.nullable();
+				} else if (Objects.equals(expectedNullability, Boolean.FALSE)) {
+					return newDataType.notNull();
+				}
+				return newDataType;
+			})
+			// preserve bridging class if possible
+			.map(newDataType -> {
+				final Class<?> clazz = actualDataType.getConversionClass();
+				final LogicalType newType = newDataType.getLogicalType();
+				if (newType.supportsInputConversion(clazz)) {
```

Comment:
Shouldn't it be `OutputConversionClass` ? Arguments of a UDF are "outputs` of the table ecosystem.

Reply:
Absolutely!